spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcloud/integration/LocalServerTestSupport.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2012-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.function.adapter.gcloud.integration;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import com.google.cloud.functions.invoker.runner.Invoker;
+import com.google.gson.Gson;
+
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.cloud.function.adapter.gcloud.FunctionInvoker;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.ResponseEntity;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Test support class for running tests on the local Cloud Function server.
+ *
+ * @author Daniel Zou
+ * @author Mike Eltsufin
+ */
+public class LocalServerTestSupport {
+
+	private static final Gson gson = new Gson();
+
+	private static final ExecutorService EXECUTOR = Executors.newCachedThreadPool();
+
+	private static final String SERVER_READY_STRING = "Started ServerConnector";
+
+	private static int nextPort = 8080;
+
+	/**
+	 * Starts up the Cloud Function Server and executes the test
+	 */
+	public static <I, O> void verify(Class mainClass, String function, I input, O expectedOutput) {
+		try (ServerProcess serverProcess = LocalServerTestSupport.startServer(mainClass, function)) {
+			TestRestTemplate testRestTemplate = new TestRestTemplate();
+
+			HttpHeaders headers = new HttpHeaders();
+
+			ResponseEntity<String> response = testRestTemplate.postForEntity(
+					"http://localhost:" + serverProcess.getPort(), new HttpEntity<>(gson.toJson(input), headers),
+					String.class);
+
+			assertThat(response.getBody()).isEqualTo(gson.toJson(expectedOutput));
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	private static ServerProcess startServer(Class<?> springApplicationMainClass, String function)
+			throws InterruptedException, IOException {
+		int port = nextPort;
```

Comment:
I think to keep it simple you can just omit updating incrementing the port if you are handling destroying the process correctly.

Reply:
@dzou what if we want to run tests in parallel?

Reply:
I don't think the parallel idea would work, the `nextPort` integer is not synchronized anyway so multithreaded execution of tests is still not guaranteed unique port with this method. 

I would just recommend against trying multi-threaded execution; we almost never do it anyway.

Reply:
This has more to do with if there is some kind of failure that causes the server not to shutdown properly or not quickly enough, holding on to the port. I'll make it thread safe.